google/cloud/bigtable/tests/instance_admin_emulator.cc

```diff
@@ -187,17 +190,25 @@ class InstanceAdminEmulator final
   grpc::Status ListClusters(grpc::ServerContext* context,
                             btadmin::ListClustersRequest const* request,
                             btadmin::ListClustersResponse* response) override {
-    std::string prefix = request->parent();
-
     // magical instance name "-" must return all clusters in the project
-    if (request->parent().back() == '-') {
-      for (auto const& kv : clusters_) {
-        *response->add_clusters() = kv.second;
+    auto project_path =
+        request->parent().substr(0, request->parent().find("/instances"));
+    auto prefix = request->parent() + "/clusters/";
+    auto magical_instance_name_found =
+        (prefix.find("/instances/-/") != std::string::npos) ? true : false;
```

Comment:
Why the ternary operator here?  `!=` already returns a boolean.

Reply:
Also, there are many ways to trick this function into returning all the instances, e.g., if I give it the "-/foo" as an instance name.  Please check for the condition you really want:

```C++
auto magical_instance_name_found = request->parent() == project_path + "/instances/-";
```

that is simple to reason about, and it is much harder to trick (I think).

Reply:
I really not thought about that.
Now changed and tested on production